### ISSUE TITLE
defgen incorrectly propagates handled Failures to the outer Deferred

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#7210]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)| @exarkun reported|
|-|-|
|Trac ID|trac#1709|
|Type|defect|
|Created|2006-05-09 13:55:49Z|

<details><summary>Searchable metadata</summary>

```
trac-id__1709 1709
type__defect defect
reporter__exarkun exarkun
priority__highest highest
milestone__ 
branch__ 
branch_author__ 
status__closed closed
resolution__fixed fixed
component__core core
keywords__ 
time__1147182949000000 1147182949000000
changetime__1147477024000000 1147477024000000
version__ 
owner__ 

```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)|@exarkun commented|
|-|-|

Ready for review in defgen-handle-error-1709

**Comment by twisted-trac** [comment]
|[<img alt="radix's avatar" src="https://avatars.githubusercontent.com/u/227068?s=50" width="50" height="50">](https://github.com/radix)|@radix commented|
|-|-|

Very happy with the reduction of code and removal of hacks. 

I will note that this changes semantics slightly more than the ticket mentions: not only does it stop propagation of Failures from handled deferreds, it also stops propagating successes. I'd bet five bucks there's code that relies on this behavior, but it is undocumented and untested behavior, so I think the branch should still be merged.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)|@exarkun set status to `closed`|
|-|-|

(In [16779]) Merge defgen-error-handling-1709

Author: exarkun
Reviewer: radix
Fixes [#7210](https://github.com/twisted/twisted/issues/7210)

This changes deferredGenerator-wrapped generator functions to only produce Deferreds
which fire with results which are *not* the result of a yield waitForDeferred-wrapped
Deferred.  The common annoying manifestation of this bug is that a generator which
handles an exception from waitForDeferred.getResult() with no subsequent yields
would cause the deferredGenerator's result to be that Failure.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)|@exarkun commented|
|-|-|

(In [16782]) Merge defgen-error-handling-1709

Author: exarkun
Reviewer: radix
Fixes [#7210](https://github.com/twisted/twisted/issues/7210)

This changes deferredGenerator-wrapped generator functions to only produce Deferreds
which fire with results which are *not* the result of a yield waitForDeferred-wrapped
Deferred.  The common annoying manifestation of this bug is that a generator which
handles an exception from waitForDeferred.getResult() with no subsequent yields
would cause the deferredGenerator's result to be that Failure.

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|
